python/tank/context.py

```diff
@@ -624,37 +623,105 @@ def _fields_from_template_tree(self, template, fields, entities):
+                        continue
+
+                    # find fields for any paths associated with this entity by looking in the path cache:
+                    entity_fields = _values_from_path_cache(entities[key_name], template, path_cache, 
+                                                           required_fields=found_fields)
+
+                    # entity_fields may contain additional fields that correspond to entities
+                    # so we should be sure to validate these as well if we can.
+                    #
+                    # The following example illustrates where the code could previously return incorrect entity 
+                    # information from this method:
+                    #
+                    # With the following template:
+                    #    /{Sequence}/{Shot}/{Step}
+                    #
+                    # And a path cache that contains:
+                    #    Type     | Id  | Name     | Path
+                    #    ----------------------------------------------------
+                    #    Sequence | 001 | Seq_001  | /Seq_001
+                    #    Shot     | 002 | Shot_A   | /Seq_001/Shot_A
+                    #    Step     | 003 | Lighting | /Seq_001/Shot_A/Lighting
+                    #    Step     | 003 | Lighting | /Seq_001/blah/Shot_B/Lighting   <- this is out of date!
+                    #    Shot     | 004 | Shot_B   | /Seq_001/blah/Shot_B            <- this is out of date!
+                    #
+                    # (Note: the schema/templates have been changed since the entries for Shot_b were added)
+                    #
+                    # The sub-templates used to search for fields are:
+                    #    /{Sequence}
+                    #    /{Sequence}/{Shot}
+                    #    /{Sequence}/{Shot}/{Step}
+                    #
+                    # And the entities passed into the method are:
+                    #    Sequence:   Seq_001
+                    #    Shot:       Shot_B
+                    #    Step:       Lighting
+                    #
+                    # We are searching for fields for 'Shot_B' that has a broken entry in the path cache so the fields 
+                    # returned for each level of the template will be:
+                    #    /{Sequence}                 -> {"Sequence":"Seq_001"} <- Correct
+                    #    /{Sequence}/{Shot}          -> {}                     <- entry not found for Shot_B matching 
+                    #                                                             the template
+                    #    /{Sequence}/{Shot}/{Step}   -> {"Sequence":"Seq_001", <- Correct
+                    #                                    "Shot":"Shot_A",      <- Wrong!
+                    #                                    "Step":"Lighting"}    <- Correct
+                    #
+                    # The final fields returned are:
```

Comment:
maybe instead say "In previous implementations, the final fields would then *incorrectly* be returned as"